zebra-consensus/src/transaction.rs

```diff
@@ -313,15 +313,17 @@ where
             check::has_inputs_and_outputs(&tx)?;
             check::has_enough_orchard_flags(&tx)?;
 
-            if req.is_mempool() && tx.has_any_coinbase_inputs() {
+            if req.is_mempool() && tx.is_coinbase() {
                 return Err(TransactionError::CoinbaseInMempool);
             }
-            if tx.has_valid_coinbase_transaction_inputs() {
+            if tx.is_coinbase() {
                 check::coinbase_tx_no_prevout_joinsplit_spend(&tx)?;
+            } else if !tx.is_valid_non_coinbase() {
+                return Err(TransactionError::NonCoinbaseHasCoinbaseInput);
```

Comment:
The previous logic had exactly the same effect, but it seemed harder to understand.